include/rmm/detail/stack_trace.hpp

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+// execinfo is a linux-only library, so stack traces will only be available on
+// linux systems.
+#if (defined(__GNUC__) && !defined(__MINGW32__) && !defined(__MINGW64__))
+#define ENABLE_STACK_TRACES
+#endif
+
+#if defined(ENABLE_STACK_TRACES)
+#include <execinfo.h>
+#include <memory>
+#include <sstream>
+#include <vector>
+#endif  // defined(ENABLE_STACK_TRACES)
+
+namespace rmm {
+
+namespace detail {
+
+/**
+ * @brief stack_trace is a class that will capture a stack on instatiation for output later.
+ * It can then be used in an output stream to display stack information.
+ *
+ * rmm::detail::stack_trace saved_stack;
+ *
+ * std::cout << "callstack: " << saved_stack;
+ *
+ */
+class stack_trace {
+ public:
+  stack_trace()
+  {
+#if defined(ENABLE_STACK_TRACES)
```

Comment:
Since nearly all of the code is inside ifdefs, why not just put ifdef around the entire #includes and the class definition, and in the else define `stack_trace` with empty function bodies.  I think this will separate the implementations much more cleanly.

Reply:
I was trying to keep a skeleton copy of the class around so other code wouldn't have to change based on if it is included or not. I agree it is messy, but thought it was the easiest way to ensure the skeleton stayed up to date since it requires a different platform to verify. I agree though that it's probably too much. I'll look into this.